packages/mdc-animation/index.ts

```diff
@@ -0,0 +1,98 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+interface VendorPropertyMap {
+  [key: string]: VendorProperty;
+}
+
+interface VendorProperty {
+  noPrefix: string;
+  styleProperty?: string;
+  webkitPrefix: string;
+}
+
+const transformStyleProperties = ['transform', 'WebkitTransform', 'MozTransform', 'OTransform', 'MSTransform'];
+
+const eventTypeMap: VendorPropertyMap = {
+  animationend: {
+    noPrefix: 'animationend',
+    styleProperty: 'animation',
+    webkitPrefix: 'webkitAnimationEnd',
+  },
+  animationiteration: {
+    noPrefix: 'animationiteration',
+    styleProperty: 'animation',
+    webkitPrefix: 'webkitAnimationIteration',
+  },
+  animationstart: {
+    noPrefix: 'animationstart',
+    styleProperty: 'animation',
+    webkitPrefix: 'webkitAnimationStart',
+  },
+  transitionend: {
+    noPrefix: 'transitionend',
+    styleProperty: 'transition',
+    webkitPrefix: 'webkitTransitionEnd',
+  },
+};
+
+const cssPropertyMap: VendorPropertyMap = {
+  animation: {
+    noPrefix: 'animation',
+    webkitPrefix: '-webkit-animation',
+  },
+  transform: {
+    noPrefix: 'transform',
+    webkitPrefix: '-webkit-transform',
+  },
+  transition: {
+    noPrefix: 'transition',
+    webkitPrefix: '-webkit-transition',
+  },
+};
+
+function isWindow(windowObj: Window): boolean {
+  return Boolean(windowObj.document) && typeof windowObj.document.createElement === 'function';
+}
+
+function getCorrectEventName(windowObj: Window, eventType: string): string {
```

Comment:
Its more accurate to make the return type of this one of the above property names, correct? What if we change those to `enums`? Then this method can return one of those.

Reply:
I'm not opposed to the idea, but I see two potential issues with enums:

1. It makes the whole file harder to read IMO (unless there's a better way to write it?)
2. We'll have to expose the enums publicly and replace all usages of `'animationend'` with `mdc.textfield.JsEventName.ANIMATION_END_NO_PREFIX` (or whatever we name it), which will make calling code more verbose

But maybe the additional verbosity is worth it for better type safety. WDYT?

E.g.:

```js
enum JsEventName {
  ANIMATION_END_NO_PREFIX = 'animationend',
  ANIMATION_END_WEBKIT_PREFIX = 'webkitAnimationEnd',

  ANIMATION_ITERATION_NO_PREFIX = 'animationiteration',
  ANIMATION_ITERATION_WEBKIT_PREFIX = 'webkitAnimationIteration',

  ANIMATION_START_NO_PREFIX = 'animationstart',
  ANIMATION_START_WEBKIT_PREFIX = 'webkitAnimationStart',

  TRANSITION_END_NO_PREFIX = 'transitionend',
  TRANSITION_END_WEBKIT_PREFIX = 'webkitTransitionEnd',
}

enum CssPropertyName {
  ANIMATION_NO_PREFIX = 'animation',
  ANIMATION_WEBKIT_PREFIX = '-webkit-animation',

  TRANSFORM_NO_PREFIX = 'transform',
  TRANSFORM_WEBKIT_PREFIX = '-webkit-transform',

  TRANSITION_NO_PREFIX = 'transition',
  TRANSITION_WEBKIT_PREFIX = '-webkit-transition',
}

interface CssVendorPropertyMap {
  [key: string]: CssVendorProperty;
}

interface JsVendorPropertyMap {
  [key: string]: JsVendorProperty;
}

interface CssVendorProperty {
  noPrefix: CssPropertyName;
  webkitPrefix: CssPropertyName;
}

interface JsVendorProperty {
  noPrefix: JsEventName;
  styleProperty: CssPropertyName;
  webkitPrefix: JsEventName;
}

const transformStyleProperties = ['transform', 'WebkitTransform', 'MozTransform', 'OTransform', 'MSTransform'];

const jsEventTypeMap: JsVendorPropertyMap = {
  [JsEventName.ANIMATION_END_NO_PREFIX]: {
    noPrefix: JsEventName.ANIMATION_END_NO_PREFIX,
    styleProperty: CssPropertyName.ANIMATION_NO_PREFIX,
    webkitPrefix: JsEventName.ANIMATION_END_WEBKIT_PREFIX,
  },
  [JsEventName.ANIMATION_ITERATION_NO_PREFIX]: {
    noPrefix: JsEventName.ANIMATION_ITERATION_NO_PREFIX,
    styleProperty: CssPropertyName.ANIMATION_NO_PREFIX,
    webkitPrefix: JsEventName.ANIMATION_ITERATION_WEBKIT_PREFIX,
  },
  [JsEventName.ANIMATION_START_NO_PREFIX]: {
    noPrefix: JsEventName.ANIMATION_START_NO_PREFIX,
    styleProperty: CssPropertyName.ANIMATION_NO_PREFIX,
    webkitPrefix: JsEventName.ANIMATION_START_WEBKIT_PREFIX,
  },
  [JsEventName.TRANSITION_END_NO_PREFIX]: {
    noPrefix: JsEventName.TRANSITION_END_NO_PREFIX,
    styleProperty: CssPropertyName.TRANSITION_NO_PREFIX,
    webkitPrefix: JsEventName.TRANSITION_END_WEBKIT_PREFIX,
  },
};

const cssPropertyNameMap: CssVendorPropertyMap = {
  [CssPropertyName.ANIMATION_NO_PREFIX]: {
    noPrefix: CssPropertyName.ANIMATION_NO_PREFIX,
    webkitPrefix: CssPropertyName.ANIMATION_WEBKIT_PREFIX,
  },
  [CssPropertyName.TRANSFORM_NO_PREFIX]: {
    noPrefix: CssPropertyName.TRANSFORM_NO_PREFIX,
    webkitPrefix: CssPropertyName.TRANSFORM_WEBKIT_PREFIX,
  },
  [CssPropertyName.TRANSITION_NO_PREFIX]: {
    noPrefix: CssPropertyName.TRANSITION_NO_PREFIX,
    webkitPrefix: CssPropertyName.TRANSITION_WEBKIT_PREFIX,
  },
};

function isWindow(windowObj: Window): boolean {
  return Boolean(windowObj.document) && typeof windowObj.document.createElement === 'function';
}

function getCorrectEventName(windowObj: Window, eventType: JsEventName): JsEventName {
  if (isWindow(windowObj) && eventType in jsEventTypeMap) {
    const el = windowObj.document.createElement('div');
    const {noPrefix, webkitPrefix, styleProperty} = jsEventTypeMap[eventType];
    const isStandard = styleProperty in el.style;
    return isStandard ? noPrefix : webkitPrefix;
  }
  return eventType;
}

function getCorrectPropertyName(windowObj: Window, cssProperty: CssPropertyName): CssPropertyName {
  if (isWindow(windowObj) && cssProperty in cssPropertyNameMap) {
    const el = windowObj.document.createElement('div');
    const {noPrefix, webkitPrefix} = cssPropertyNameMap[cssProperty];
    const isStandard = noPrefix in el.style;
    return isStandard ? noPrefix : webkitPrefix;
  }
  return cssProperty;
}

export {transformStyleProperties, getCorrectEventName, getCorrectPropertyName};
```

Reply:
I think it is a bit harder to read, but its not impossible. It is at least in the same file and easy to follow backwards. I think having those types will allow for quicker development when a developer uses `getCorrectEventName` or `getCorrectPropertyName` (when thinking of autocomplete).